lib/active_merchant/billing/gateways/nuvei.rb

```diff
@@ -0,0 +1,298 @@
+      self.default_currency = 'USD'
+      self.money_format = :cents
+      self.supported_cardtypes = %i[visa master american_express discover union_pay]
+      self.currencies_without_fractions = %w[CLP KRW JPY ISK MMK PYG UGX VND XAF XOF]
+      self.homepage_url = 'https://www.nuvei.com/'
+      self.display_name = 'Nuvei'
+
+      ENDPOINTS_MAPPING = {
+        authenticate: '/getSessionToken',
+        purchase: '/payment', # /authorize with transactionType: "Auth"
+        capture: '/settleTransaction',
+        refund: '/refundTransaction',
+        void: '/voidTransaction',
+        general_credit: '/payout'
+      }
+
+      def initialize(options = {})
+        requires!(options, :merchant_id, :merchant_site_id, :secret_key)
+        super
+      end
+
+      def authorize(money, payment, options = {})
+        post = {}
+        post[:transactionType] = 'Auth'
+        build_post_data(post, :authorize)
+        add_invoice(post, money, options)
+        add_payment(post, payment)
+        add_address(post, payment, options)
+        add_customer_ip(post, options)
+
+        commit(:purchase, post)
+      end
+
+      def purchase(money, payment, options = {})
+        post = {}
+        build_post_data(post, :payment)
+        add_invoice(post, money, options)
+        add_payment(post, payment)
+        add_address(post, payment, options, :billingAddress)
+        add_address(post, payment, options, :shippingAddress)
+        add_customer_ip(post, options)
+
+        commit(:purchase, post)
+      end
+
+      def capture(money, authorization, options = {})
+        post = {}
+        post[:relatedTransactionId] = authorization
+        build_post_data(post, :capture)
+        add_invoice(post, money, options)
+
+        commit(:capture, post)
+      end
+
+      def refund(money, authorization, options = {})
+        post = {}
+        post[:relatedTransactionId] = authorization
+        build_post_data(post, :refund)
+        add_invoice(post, money, options)
+
+        commit(:refund, post)
+      end
+
+      def void(authorization, options = {})
+        post = {}
+        post[:relatedTransactionId] = authorization
+        build_post_data(post, :void)
+
+        commit(:void, post)
+      end
+
+      def verify(credit_card, options = {})
+        MultiResponse.run(:use_first_response) do |r|
+          r.process { authorize(0, credit_card, options) }
+          r.process(:ignore_result) { void(r.authorization, options) }
+        end
+      end
+
+      def credit(money, payment, options = {})
+        post = {}
+        post[:userTokenId] = options[:user_token_id] if options[:user_token_id]
+        build_post_data(post, :general_credit)
+        add_invoice(post, money, options)
+        add_payment(post, payment, :cardData)
+        add_address(post, payment, options)
+        add_customer_ip(post, options)
+        # The card details should be nested under cardData
+        card_details = post[:cardData].delete(:card)
```

Comment:
💭 Opinion/Suggestion

I think that the add_payment could be improved to avoid those object modifications. I'm not sure if something like this:

```ruby
  def add_payment(post, payment, key = :paymentOption)
  payment_data = if payment.is_a?(CreditCard)
                   credit_card_hash(payment)
                 else
                   { cardToken: payment }
                 end

  if key == :cardData
    post[key] = payment_data
  else
    post[key] = { card: payment_data }
  end
end
```
Or something different.

Reply:
yes, probably would be another code refactoring here.